internal/key/store.go

```diff
@@ -60,3 +68,80 @@ type Iterator interface {
 	// while iterating over the set of keys.
 	Err() error
 }
+
+// StoreState describes the state of a Store.
+type StoreState struct {
+	// State is the state of the Store. A Store
+	// can either be reachable or unreachable.
+	State string
+
+	// Latency is the time it took to reach
+	// the Store.
+	Latency time.Duration
+}
+
+const (
+	StoreReachable   = "reachable"
+	StoreUnreachable = "unreachable"
+)
+
+// DialStore dials to the Store at the given endpoint
+// and returns a StoreState describing the Store status.
+//
+// If endpoint does not contain any URL scheme, DialStore
+// uses the https URL scheme as default.
+func DialStore(ctx context.Context, endpoint string) (StoreState, error) {
+	const (
+		HTTPS       = "https://"
+		DefaultPort = "443"
+	)
+
+	URL, err := url.Parse(endpoint)
+	if err != nil {
+		return StoreState{}, err
+	}
+	if URL.Hostname() == "" {
+		// If the URL does not contain a hostname
+		// the raw endpoint does not contain a
+		// scheme. For example: localhost:443
+		// instead of https://localhost:443.
+		//
+		// In this case, we prepend the
+		// https:// scheme to obtain the
+		// hostname and port later on.
+		URL, err = url.Parse(HTTPS + endpoint)
+		if err != nil {
+			return StoreState{}, err
+		}
+	}
+
+	var (
+		host = URL.Hostname()
+		port = URL.Port()
+	)
+	if port == "" {
+		port = DefaultPort
+	}
+
+	var (
+		d     net.Dialer
+		start = time.Now()
+	)
+	c, err := d.DialContext(ctx, "tcp", host+":"+port)
```

Comment:
```suggestion
	c, err := d.DialContext(ctx, "tcp", net.JoinHostPort(host, port))
```

Reply:
I think it's better to also reach out to https://www.vaultproject.io/api/system/health for Vault-specific endpoints to health. 

Generic Dial is okay and serves it purpose but we should consider writing health checks for individual targets.

Reply:
> Generic Dial is okay and serves it purpose but we should consider writing health checks for individual targets.

Yes, that we can do. I added a health check for Vault. Others can follow. However, I don't want to expose target-specific details in the API. For example, AWS is not in a sealed or not initialized state - in contrast to Vault.

For a client it is sufficient to distinguish:
 - `available`: The KMS is reachable and ready to serve requests
 - `reachable`: The KMS is reachable but is may not be able to serve requests
 - `unreachable`: The KMS is not reachable (over the network)